### PR TITLE
Adds config options for async callbacks.

### PR DESCRIPTION
**Pull request by samjonester**
To better represent the asynchronous nature of JavaScript development,
this commit adds two new configuration options for stubbing methods that
accept callbacks.

1. {delay: 10} - the delay option will delay invocation of the
callback provided to a test double by the provided number of
milliseconds
2. {defer: true} - the defer option will defer invocation of the
callback provided to a occur after the current callstack has cleared.

This closes: Support invoking callbacks with arbitrary timing #106

### CODE REVIEW

**Comment by samjonester** [comment]
Depending on whether this API is acceptable, documentation will need to be added.

As an alternative API, `td.when().thenReturnAsync()` + `td.when().thenCallbackAsync()` might be more discoverable. They also might prevent confusion about when these new config options are valid.

File: test/src/callback-test.coffee [hunk 1]
```diff
@@ -2,52 +2,87 @@ describe 'td.callback', ->
...
+        And -> @results == undefined
+        And -> @callbackInvoked == true
+
+      context 'EDGE CASE: thenCallback used but not satisfied', ->
+        Given -> td.when(@testDouble('/bar')).thenCallback('a-ha')
+        Given -> td.when(@testDouble('/bar')).thenReturn('o_O')
+        When -> @result = @testDouble('/bar')
+        Then -> @result == 'o_O'
+
+    context 'callback is asynchronous', ->
```
Comment (samjonester) [inline test/src/callback-test.coffee]: I couldn't think of a cleaner way to represent the branching in this test. I'm up for suggestions.

File: src/store/stubbings.coffee [hunk 2]
```diff
@@ -42,15 +42,24 @@ invokeCallbackFor = (stubbing, actualArgs) ->
   return unless _.some(stubbing.args, callback.isCallback)
   _.each stubbing.args, (expectedArg, i) ->
     return unless callback.isCallback(expectedArg)
-    callbackArgs = if expectedArg.args?
-      expectedArg.args
-    else if stubbing.config.plan == 'thenCallback'
-      stubbing.stubbedValues
-    else
-      []
+    args = callbackArgs stubbing, expectedArg
```
Comment (searls) [inline src/store/stubbings.coffee]: if you wouldn't mind, all the single-line function invocations use `()`s in this project

File: test/src/callback-test.coffee (hunk 1 above)
Comment (searls) [inline test/src/callback-test.coffee]: Branching in this way is fine given that there's a shared `when` tying your hands at the top.

File: test/src/callback-test.coffee [hunk 3]
```diff
@@ -2,52 +2,87 @@ describe 'td.callback', ->
...
+      When (done) -> @returnValue = @testDouble '/foo', (er, results) =>
+        @callbackInvoked = true
+        @er = er
+        @results = results
+        done()
```
Comment (searls) [inline test/src/callback-test.coffee]: I don't think these tests do anything to ensure that the result is actually deferred, do they? What if you added a line at the end to make sure the results weren't set yet:

```coffee
     When (done) -> @returnValue = @testDouble '/foo', (er, results) =>
          @callbackInvoked = true
          @er = er
          @results = results
          done()
        @invokedSynchronously = @results?
    Invariant -> @invokedSynchronously == false
```

Maybe that'd do the trick?

Comment (samjonester) [inline test/src/callback-test.coffee]: Seems like valuable addition. Definitely worth a try!

File: src/store/stubbings.coffee (hunk 2 above)
Comment (samjonester) [inline src/store/stubbings.coffee]: +1
